website/src/utils/timetables.ts

```diff
@@ -434,18 +443,68 @@ export function formatNumericWeeks(weeks: NumericWeeks): string | null {
 // Converts a timetable config to query string
 // eg:
 // {
-//   CS2104: { Lecture: '1', Tutorial: '2' },
-//   CS2107: { Lecture: '1', Tutorial: '8' },
+//   CS2104: { Lecture: '1', Tutorial: '2', Color: 3 },
+//   CS2107: { Lecture: '1', Tutorial: '8', Color: 4 },
 // }
-// => CS2104=LEC:1,Tut:2&CS2107=LEC:1,Tut:8
-export function serializeTimetable(timetable: SemTimetableConfig): string {
+// => CS2104=LEC:1,TUT:2,COL:3&CS2107=LEC:1,TUT:8,COL:4
+export function serializeTimetable(
+  timetable: SemTimetableConfig,
+  colors?: ColorMapping | null,
+  themeId?: ThemeId | null,
+): string {
+  const timetableModuleParams = Object.fromEntries(
+    Object.entries(timetable).map((module) => {
+      const [moduleCode, moduleLessonConfig] = module;
+      const moduleParams: { [param: string]: string | ColorIndex } = { ...moduleLessonConfig };
+      if (colors && moduleCode in colors) {
+        moduleParams[COLOR_PARAM] = colors[moduleCode];
+      }
+      return [moduleCode, moduleParams];
+    }),
+  );
+  const timetableParams = {
+    ...mapValues(timetableModuleParams, serializeModuleConfig),
+    ...((themeId && { [THEME_PARAM]: themeId }) || EMPTY_OBJECT),
+  };
```

Comment:
This is a bit easier to read, but also use `?:` ternary to represent ternary logic instead of `&& ||`

```suggestion
  const timetableParams = mapValues(timetableModuleParams, serializeModuleConfig);
  if (themeId != null) {
    timetableParams[THEME_PARAM] = themeId };
  }
```